### PR TITLE
add master slave heartbeat and kill invalid thread id

### PR DESCRIPTION
**Pull request by anoty**
When the network jitter or other reasons, the mysql server does not kill invalid canal database connection, which can cause waste of resources, at the same time, with the heartbeat set, canal initiative to kill invalid connection, and to connect to the database

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @WangXiangUSTC

File: cmd/go-canal/main.go [hunk 1]
```diff
@@ -30,6 +31,9 @@ var ignoreTables = flag.String("ignore_tables", "", "ignore tables, must be data
 var startName = flag.String("bin_name", "", "start sync from binlog name")
 var startPos = flag.Uint("bin_pos", 0, "start sync from binlog position of")
 
+var heartbeatPeriod = flag.Uint("heartbeat", 60, "master heartbeat period")
```
Comment (siddontang) [inline cmd/go-canal/main.go]: should use flag.Duration

File: cmd/go-canal/main.go [hunk 2]
```diff
@@ -30,6 +31,9 @@ var ignoreTables = flag.String("ignore_tables", "", "ignore tables, must be data
...
+var readTimeout = flag.Uint("read_timeout", 90, "connection read timeout")
```
Comment (siddontang) [inline cmd/go-canal/main.go]: ditto

File: replication/binlogsyncer.go [hunk 3]
```diff
@@ -62,6 +63,12 @@ type BinlogSyncerConfig struct {
 
 	// RecvBufferSize sets the size in bytes of the operating system's receive buffer associated with the connection.
 	RecvBufferSize int
+
+	// master heartbeat period
+	HeartbeatPeriod uint
```
Comment (siddontang) [inline replication/binlogsyncer.go]: use time.Duration

File: replication/binlogsyncer.go [hunk 4]
```diff
@@ -163,16 +172,38 @@ func (b *BinlogSyncer) registerSlave() error {
...
 		}
 	}
 
+	// kill last connection id
+	if b.lastConnectionID > 0 {
+		cmd := fmt.Sprintf("KILL %d", b.lastConnectionID)
+		if _, err := b.c.Execute(cmd); err != nil {
+			log.Errorf("b.c.Execute(%s) error(%v)", cmd, err)
+			// unknown connection id
+			if !strings.Contains(err.Error(), "ERROR 1094") {
```
Comment (siddontang) [inline replication/binlogsyncer.go]: can we get the error code instead of the error message directly?

**Comment by anoty** [comment]
@siddontang All right, please check it again

File: replication/binlogsyncer.go [hunk 5]
```diff
@@ -163,16 +171,42 @@ func (b *BinlogSyncer) registerSlave() error {
...
 		if tcp, ok := b.c.Conn.Conn.(*net.TCPConn); ok {
 			tcp.SetReadBuffer(b.cfg.RecvBufferSize)
 		}
 	}
 
+	// kill last connection id
+	if b.lastConnectionID > 0 {
+		cmd := fmt.Sprintf("KILL %d", b.lastConnectionID)
+		if _, err := b.c.Execute(cmd); err != nil {
+			log.Errorf("b.c.Execute(%s) error(%v)", cmd, err)
```
Comment (siddontang) [inline replication/binlogsyncer.go]:
```
log.Errrof("kill connection %d error %v", b.lastConnectionID, err) 
```

File: replication/binlogsyncer.go [hunk 6]
```diff
@@ -163,16 +171,42 @@ func (b *BinlogSyncer) registerSlave() error {
...
+			var tmpStr string
+			var code int
+			// golang scanf doesn't support %*,so I used a temporary variable
+			fmt.Sscanf(err.Error(), "%s%d", &tmpStr, &code)
```
Comment (siddontang) [inline replication/binlogsyncer.go]: can we ensure the message style is string + ID here? If not, I worry the scan may panic here. 
If we can't guarantee this, we may still check error message.

**Comment by siddontang** [comment]
any test to check this feature?

File: replication/binlogsyncer.go (hunk 6 above)
Comment (anoty) [inline replication/binlogsyncer.go]: I've tested it here, and if the string isn't in the right format, code is 0, not panic

**Comment by anoty** [comment]
@siddontang Tests have been added for the feature

File: canal/canal_test.go [hunk 7]
```diff
@@ -57,6 +62,22 @@ func (s *canalTestSuite) SetUpSuite(c *C) {
 }
 
 func (s *canalTestSuite) TearDownSuite(c *C) {
+	sc := make(chan os.Signal, 1)
```
Comment (siddontang) [inline canal/canal_test.go]: I am confused with these, why not sleep directly but introduce complex signal and channel?

File: replication/binlogsyncer.go (hunk 6 above)
Comment (siddontang) [inline replication/binlogsyncer.go]: I prefer extracting a simple function like extractErrorCode, so we can test it directly. 

```
func extractErrorCode(err Error) int {
   var (
      tmpStr string 
      code int 
   )

   fmt.Sscanf...
   return code 
}
```

Comment (anoty) [inline replication/binlogsyncer.go]: I'm going to take the extraction method and put it in [error.go](https://github.com/siddontang/go-mysql/blob/master/mysql/error.go)，Is that ok？

File: canal/canal_test.go (hunk 7 above)
Comment (anoty) [inline canal/canal_test.go]: Oh，you are right！

**Comment by WangXiangUSTC** [comment]
nanosecond := b.cfg.HeartbeatPeriod * 1000000000
 _, err = b.c.Execute(fmt.Sprintf("SET @master_heartbeat_period =%d;", nanosecond))
can you confirm the master_heartbeat_period is set use ns? @anoty

**Comment by anoty** [comment]
@WangXiangUSTC I am sure, and I used wireshark to grab the packet to confirm that it was a nanosecond，Mysql 5.7 master_heartbeat_period  defaults to 30s，you can try it with Wireshark.

File: replication/binlogsyncer.go (hunk 6 above)
Comment (siddontang) [inline replication/binlogsyncer.go]: cool

File: mysql/mysql_test.go [hunk 8]
```diff
@@ -151,3 +151,17 @@ func (t *mysqlTestSuite) TestMysqlParseBinaryUint64(c *check.C) {
 	u64 := ParseBinaryUint64([]byte{1, 2, 3, 4, 5, 6, 7, 128})
 	c.Assert(u64, check.Equals, 128*uint64(72057594037927936)+7*uint64(281474976710656)+6*uint64(1099511627776)+5*uint64(4294967296)+4*16777216+3*65536+2*256+1)
 }
+
+func (t *mysqlTestSuite) TestErrorCode(c *check.C) {
+	var code int
```
Comment (siddontang) [inline mysql/mysql_test.go]: you can use table-driven test instead like 
```
tbl := []struct {
    msg string 
    code int 
} {
     {"xxx", xxx},
}

for t := range tbls {
   c.Assert(ErrorCode(t.msg), check.Equals, t.code)
}
```

File: canal/canal_test.go [hunk 9]
```diff
@@ -28,6 +28,8 @@ func (s *canalTestSuite) SetUpSuite(c *C) {
 	cfg := NewDefaultConfig()
 	cfg.Addr = fmt.Sprintf("%s:3306", *testHost)
 	cfg.User = "root"
+	cfg.HeartbeatPeriod = 2
```
Comment (siddontang) [inline canal/canal_test.go]: period and timeout are duration, so here is 2 and 3 nanosecond.

Comment (siddontang) [inline canal/canal_test.go]: I prefer using 200ms and 300ms instead, so the latter sleep can only 1s, no need 4s.

**Comment by siddontang** [comment]
PTAL @WangXiangUSTC

**Comment by anoty** [comment]
@siddontang done，PTAL

File: cmd/go-canal/main.go [hunk 10]
```diff
@@ -39,6 +43,8 @@ func main() {
 	cfg.Password = *password
 	cfg.Flavor = *flavor
 
+	cfg.ReadTimeout = *readTimeout * time.Second
```
Comment (siddontang) [inline cmd/go-canal/main.go]: no need to multi time.Second 
you should use "60s" instead of 60 in the command flag.

Comment (anoty) [inline cmd/go-canal/main.go]: flag.Duration does not support the string！Are there any other way？

**Comment by anoty** [comment]
PTAL @siddontang

**Pull request merged**